### PR TITLE
make linter only run on changed files 

### LINKED ISSUE #392: Improve linter phase speed in CI
Everything in CI runs in 2 minutes except the linter phase takes more than 25 minutes. This slows down the reaction time, e.g. notification messages come up, one visits the ticket but doesn't review since the CI hasn't given the OK yet. So one then comes back next day or so when scanning GitHub messages. Or the review was done but the author can't merge yet.

### PR DESCRIPTION
**Pull request by Taepper**
resolves #392 #491 

### Summary

Now for every pull request the linter is only run on the files, that are changed by the pull request.

We will also add a check for all files again that is run on main and on demand.

## PR Checklist

~~- [ ] All necessary documentation has been adapted or there is an issue to do so.~~
~~- [ ] The implemented feature is covered by an appropriate test.~~

### CODE REVIEW

**Comment by fengelniederhammer** [comment]
Maybe an overall thought on this: Is it a good idea to only consider changed files?

Are there rules that - given I changed file A - would fail in file B?

File: .github/workflows/linter.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+          echo "Github Event ${{ github.event }}"
+          echo "Github Event number ${{ github.event.number }}"
+          echo "github.repository_owner ${{ github.repository_owner }}"
+          echo "github.repository ${{ github.repository }}"
+          echo "Changed files in PR #${PR_NUMBER}:"
+          $files=$(curl -s \
+           "https://api.github.com/repos/${{ github.repository }}/pulls/${PR_NUMBER}/files" \
+           | jq -r '.[].filename')
+          echo "${$files}"
+          echo "::set-output name=files::$files"
```
Comment (fengelniederhammer) [inline .github/workflows/linter.yml:35]: Is this step actually necessary?

File: .github/workflows/linter.yml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+          cache-from: type=gha,ref=linter-dependencies-image-cache-${{ hashFiles('conanfile.py', 'Dockerfile_linter') }}
+          cache-to: type=gha,mode=min,ref=linter-dependencies-image-cache-${{ hashFiles('conanfile.py', 'Dockerfile_linter') }}
+
+      - name: Retag and push existing image if cache hit
+        if: env.CACHE_HIT == 'true'
+        run: |
+          TAGS=(${{ steps.dockerMetadata.outputs.tags }})
+          for TAG in "${TAGS[@]}"; do
+            docker buildx imagetools create --tag $TAG ${{ env.DOCKER_LINTER_DEPENDENCY_IMAGE_NAME }}:${{ env.DIR_HASH }}
+          done
```
Comment (fengelniederhammer) [inline .github/workflows/linter.yml:88]: I wonder whether this is actually necessary. Wouldn't it be enough to run this if `cache hit && is main branch` and then only apply the `latest` tag?

File: src/silo/query_engine/actions/action.cpp [hunk 3]
```diff
@@ -128,7 +128,7 @@ void from_json(const nlohmann::json& json, OrderByField& field) {
    CHECK_SILO_QUERY(
       json.is_object() && json.contains("field") && json.contains("order") &&
          json["field"].is_string() && json["order"].is_string(),
-      "The orderByField '" + json.dump() +
```
Comment (fengelniederhammer) [inline src/silo/query_engine/actions/action.cpp:131]: Reminder: revert before merge

File: .github/workflows/linter.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+  linter:
+    name: Build And Run linter
+    needs: linterDependencies
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/genspectrum/lapis-silo-linter-dependencies:commit-${{ github.sha }}
```
Comment (fengelniederhammer) [inline .github/workflows/linter.yml:95]: This makes the `Dockerfile_linter` superfluous, doesn't it?

File: .github/workflows/linter.yml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+           "https://api.github.com/repos/${{ github.repository_owner }}/${{ github.repository }}/pulls/${PR_NUMBER}/files" \
+           | jq -r '.[].filename')
+          echo "${$files}"
+          echo "::set-output name=files::$files"
+
+      - name: Get list of changed files
+        id: changed-files2
+        run: |
+          files=$(jq -r '.pull_request.files[] | select(.status == "added" or .status == "modified") | .filename' $GITHUB_EVENT_PATH)
+          echo "::set-output name=files::$files"
```
Comment (fengelniederhammer) [inline .github/workflows/linter.yml:116]: This seems unused

File: .github/workflows/linter.yml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+on:
+  pull_request:
+
+concurrency:
+  group: linter-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  DOCKER_LINTER_DEPENDENCY_IMAGE_NAME: ghcr.io/genspectrum/lapis-silo-linter-dependencies
```
Comment (fengelniederhammer) [inline .github/workflows/linter.yml:11]: The corresponding line in `ci.yml` can be deleted

**Review by fengelniederhammer — CHANGES_REQUESTED** [review]

File: .github/workflows/linter.yml (hunk 5 above)
Comment (Taepper) [reply .github/workflows/linter.yml:116]: Yes, I am still very much in trial and error :D

File: .github/workflows/linter.yml (hunk 4 above)
Comment (Taepper) [reply .github/workflows/linter.yml:82]: As discussed I added #498

File: .github/workflows/linter.yml (hunk 2 above)
Comment (Taepper) [reply .github/workflows/linter.yml:73]: Hmm, not quite, as we want the commit tags as well in some cases. Retagging all that would have been tagged if the cache were not hit is good practice / less complexity nonetheless I would say

File: .github/workflows/linter.yml (hunk 1 above)
Comment (Taepper) [reply .github/workflows/linter.yml]: Nope, it was just for testing. And I put it first in the file for quicker trial and error :D

**Comment by Taepper** [comment]
Okay, I got it working now, see workflow 
https://github.com/GenSpectrum/LAPIS-SILO/actions/runs/9585385926/job/26431156423
for successful linting

File: CMakeLists.txt [hunk 7]
```diff
@@ -60,24 +60,18 @@ file(GLOB_RECURSE SRC_SILO "src/*.cpp")
 list(REMOVE_ITEM SRC_SILO ${SRC_TEST})
 
 # ---------------------------------------------------------------------------
-# Linter
+# Targets
 # ---------------------------------------------------------------------------
 
-option(BUILD_WITH_CLANG_TIDY "Build process clang-tidy")
```
Comment (JonasKellerer) [inline CMakeLists.txt:66]: Can you also change the `build_with_conan.py` file to stop using this variable, but using `-D CMAKE_CXX_CLANG_TIDY=clang-tidy`? 
I tried `cmake_options.append("-D CMAKE_CXX_CLANG_TIDY=clang-tidy")` myself, but I get a lot of errors. Some dont seem to find std packages...
```
/fasta_format_exception.h:3:10: error: 'stdexcept' file not found [clang-diagnostic-error]
    3 | #include <stdexcept>
```

Comment (Taepper) [reply CMakeLists.txt]: This should be resolved as I took another approach which left all relevant files for manual linting unchanged

**Pull request merged**